lib/Sema/CSBindings.cpp

```diff
@@ -447,7 +447,8 @@ ConstraintSystem::getPotentialBindings(TypeVariableType *typeVar) {
 
     // Make sure we aren't trying to equate type variables with different
     // lvalue-binding rules.
-    if (auto otherTypeVar = type->getAs<TypeVariableType>()) {
+    if (auto otherTypeVar =
+            type->lookThroughAllOptionalTypes()->getAs<TypeVariableType>()) {
       if (typeVar->getImpl().canBindToLValue() !=
           otherTypeVar->getImpl().canBindToLValue())
         continue;
```

Comment:
Interesting. So presumably we were skipping this inner test because we weren't trying to bind a typevar, but an optional of a TV. That TV can eventually be bound to something of different Lvalue-ness than `typeVar`, resulting in the optional itself having the same Lvalue-ness of the nested type.

Reply:
Yes, that’s exactly one of the scenarios how this fails, another one is - bind constraint just sticks forever unsolved because there is a fallback in match types to form unsolved constraint if the lvalueness is different, which means if there are no other bindings available the system as a whole is unsolvable.